cmd/agent/app/server.go

```diff
@@ -97,15 +100,21 @@ func (a *Agent) runProxyConnection(o *options.GrpcProxyAgentOptions, stopCh <-ch
 	cs := cc.NewAgentClientSet(stopCh)
 	cs.Serve()
 
-	return nil
+	return cs, nil
 }
 
-func (a *Agent) runHealthServer(o *options.GrpcProxyAgentOptions) error {
+func (a *Agent) runHealthServer(o *options.GrpcProxyAgentOptions, cs agent.ReadinessManager) error {
 	livenessHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		fmt.Fprintf(w, "ok")
 	})
 	readinessHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		fmt.Fprintf(w, "ok")
+		if cs.Ready() {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprintf(w, "ok")
```

Comment:
I generally like the change. However I'm a little conerned about this as its the actual external facing API surface. I would suggest we break this down as an overall ready/not ready but also we have named checks as per [https://github.com/kubernetes/kubernetes/blob/589d6f388680b29f7f7961bda4f4262733b54210/test/e2e/apimachinery/health_handlers.go#L76](https://www.google.com/url?q=https://github.com/kubernetes/kubernetes/blob/589d6f388680b29f7f7961bda4f4262733b54210/test/e2e/apimachinery/health_handlers.go%23L76&sa=D&source=docs&ust=1686640049965118&usg=AOvVaw1biHPmKn7GLxXClPAsjSlj). Ready here is really checking whether there is an attached KonnServer. That should e a named check. I think for you purposes that is what you want to look at. While it would feed into the overall readiness, I think readiness should include more that that,.